### PR TITLE
Support vector write mask registers on x86

### PR DESCRIPTION
**Pull request by BradleyWood**
Goals:

1. Introduce new register kind: TR_VMR for vector mask registers
2. Enable write masking and merge masking operations on x86 with AVX512

### CODE REVIEW

**Comment by BradleyWood** [comment]
Jenkins build all

**Comment by BradleyWood** [comment]
Jenkins build all

**Comment by BradleyWood** [comment]
Jenkins build all

**Comment by BradleyWood** [comment]
Jenkins build all

File: compiler/x/codegen/OMRMachine.cpp [hunk 1]
```diff
@@ -445,10 +453,11 @@ TR::RealRegister *OMR::X86::Machine::freeBestGPRegister(TR::Instruction
...
 
    bool useRegisterInterferences = self()->cg()->enableRegisterInterferences() ? true : false;
-   bool enableRematerialisation = self()->cg()->enableRematerialisation() ? true : false;
+   bool enableRematerialisation = self()->cg()->enableRematerialisation() && virtReg->getKind() != TR_VMR;
+   // rematerialisation not supported for vector masks
 
-   TR_ASSERT(virtReg && (virtReg->getKind() == TR_GPR || virtReg->getKind() == TR_FPR || virtReg->getKind() == TR_VRF),
-          "OMR::X86::Machine::freeBestGPRegister() ==> expecting to free GPRs or XMMRs only!");
+   TR_ASSERT_FATAL(virtReg && (virtReg->getKind() == TR_GPR || virtReg->getKind() == TR_FPR || virtReg->getKind() == TR_VRF || virtReg->getKind() == TR_VMR),
+          "OMR::X86::Machine::freeBestGPRegister() ==> expecting to free GPRs, VMRS or XMMRs only!");
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRMachine.cpp:460]: Lower case "S" on VMRS.

File: compiler/x/codegen/OMRMachine.cpp [hunk 2]
```diff
@@ -1073,6 +1122,31 @@ TR::RealRegister *OMR::X86::Machine::reverseGPRSpillState(TR::Instruction     *c
          spilledRegister->setBackingStorage(NULL);
          }
       }
+   else if (spilledRegister->getKind() == TR_VMR)
+      {
+      // TODO
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRMachine.cpp:1127]: What's the TODO?

File: compiler/x/codegen/OMRMachine.cpp [hunk 3]
```diff
@@ -1073,6 +1122,31 @@ TR::RealRegister *OMR::X86::Machine::reverseGPRSpillState(TR::Instruction     *c
...
+         opcode = TR::InstOpCode::KMOVQMemMask;
+         spillSize = 8;
+         }
+
+      instr = new (self()->cg()->trHeapMemory())
+                TR::X86MemRegInstruction(
+                currentInstruction,
+            opcode,
+            tempMR,
+           targetRegister, self()->cg());
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRMachine.cpp:1142]: Fix indentation here.

File: compiler/x/codegen/OMRX86Instruction.hpp [hunk 4]
```diff
@@ -1361,6 +1361,135 @@ class X86RegRegRegInstruction : public TR::X86RegRegInstruction
...
+   *    The address after last operand byte
+   */
+   virtual uint8_t* generateOperand(uint8_t* cursor);
+   virtual void assignRegisters(TR_RegisterKinds kindsToBeAssigned);
+   virtual bool refsRegister(TR::Register *reg);
+   virtual bool defsRegister(TR::Register *reg);
+   virtual bool usesRegister(TR::Register *reg);
+
+#ifdef DEBUG
+        virtual uint32_t getNumOperandReferencedGPRegisters() { return 3; }
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86Instruction.hpp:1490]: Should this be "4" if you're considering the mask register to be a "GP" in this context?  If not, then the return value on L1424 needs to be reconsidered.

File: compiler/x/codegen/OMRX86Instruction.hpp [hunk 5]
```diff
@@ -2133,6 +2315,82 @@ class X86RegRegMemInstruction : public TR::X86RegMemInstruction
...
+   virtual Kind getKind() { return IsRegMaskMem; }
+
+   /** \brief
+   *    Getter of the write mask register
+   *
+   *  \return
+   *    The write mask register
+   */
+   virtual TR::Register *getMaskRegister()          {return _maskRegister;}
+   /** \brief
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86Instruction.hpp:2363]: Newline before to space things out.

File: compiler/x/codegen/OMRX86Instruction.hpp [hunk 6]
```diff
@@ -2133,6 +2315,82 @@ class X86RegRegMemInstruction : public TR::X86RegMemInstruction
...
+   *    Setter of the write mask register
+   *
+   *  \param sr
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRX86Instruction.hpp:2366]: Should be "mr" to match the actual parameter.

File: compiler/x/i386/codegen/OMRRealRegister.cpp [hunk 7]
```diff
@@ -56,4 +56,12 @@ const OMR::X86::RealRegister::TR_RegisterBinaryEncoding OMR::X86::I386::RealRegi
    {0x05, 0, 0},        // xmm6
    {0x06, 0, 0},        // xmm6
    {0x07, 0, 0},        // xmm7
+   {0x00},              // k1
```
Comment (0xdaryl) [inline compiler/x/i386/codegen/OMRRealRegister.cpp:59]: k0 ?

File: compiler/x/i386/codegen/OMRRealRegister.hpp [hunk 8]
```diff
@@ -147,6 +147,34 @@ class OMR_EXTENSIBLE RealRegister : public OMR::X86::RealRegister
...
+         case OMR::RealRegister::k4:
+            return OMR::RealRegister::k4Mask;
+         case OMR::RealRegister::k5:
+            return OMR::RealRegister::k5Mask;
+         case OMR::RealRegister::k6:
+            return OMR::RealRegister::k6Mask;
+         case OMR::RealRegister::k7:
+            return OMR::RealRegister::k7Mask;
+         default:
+            TR_ASSERT_FATAL(0, "vector mask mask valid for k1-k7 only");
```
Comment (0xdaryl) [inline compiler/x/i386/codegen/OMRRealRegister.hpp:173]: If true, should the `k0` case above assert then?

File: compiler/x/i386/codegen/OMRRealRegister.hpp [hunk 9]
```diff
@@ -182,6 +210,12 @@ class OMR_EXTENSIBLE RealRegister : public OMR::X86::RealRegister
       *opcodeByte |= _fullRegisterBinaryEncodings[_registerNumber].id; // reg field is in bits 0-2 of opcode
       }
 
+   void setMaskRegisterInEvex(uint8_t *evex, bool zero = false)
+      {
+      uint8_t regNum = getRegisterNumber() - k1 + 1;
```
Comment (0xdaryl) [inline compiler/x/i386/codegen/OMRRealRegister.hpp:215]: Suggest `OMR::RealRegister::k1` for clarity.

**Comment by BradleyWood** [comment]
Rebased...

File: compiler/x/i386/codegen/OMRRealRegister.hpp (hunk 8 above)
Comment (BradleyWood) [reply compiler/x/i386/codegen/OMRRealRegister.hpp:173]: The comment should say k0-k7

**Comment by BradleyWood** [comment]
Rebasing...

**Comment by 0xdaryl** [comment]
Jenkins build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**